File: interpreter.ml
exception RuntimeError of string;;
exception TypeError of Lang.st * Lang.st;;

let rec replace e nme v = match e with
  | Lang.Lit _ -> e
  | Lang.Add (lhs, rhs) -> Lang.Add ((replace lhs nme v), (replace rhs nme v))
  | Lang.Sub (lhs, rhs) -> Lang.Sub ((replace lhs nme v), (replace rhs nme v))
  | Lang.Les (lhs, rhs) -> Lang.Les ((replace lhs nme v), (replace rhs nme v))
  | Lang.Eqn (lhs, rhs) -> Lang.Eqn ((replace lhs nme v), (replace rhs nme v))
  | Lang.Var nme' ->
    if nme = nme' then v
    else e
  | Lang.Fun (nme', body) ->
    if nme = nme' then e
    else Lang.Fun (nme', (replace body nme v))
  | Lang.App (lhs, rhs) -> Lang.App ((replace lhs nme v), (replace rhs nme v))
  | Lang.TST (cond, csq, alt) ->
    Lang.TST ((replace cond nme v), (replace csq nme v), (replace alt nme v))
;;

let rec eval e = match e with
  | Lang.Lit n -> n
  | Lang.Add (lhs, rhs) -> (eval lhs) + (eval rhs)
  | Lang.Sub (lhs, rhs) -> (eval lhs) - (eval rhs)
  | Lang.Var nme -> raise (RuntimeError ("unbound variable: " ^ nme))
  | Lang.App (Lang.Fun (nme, body), p) -> (eval (replace body nme p))
  | Lang.TST (Lang.Les (lhs, rhs), csq, alt) ->
    let lhs' = eval lhs in let rhs' = eval rhs in
      if (lhs' < rhs') then (eval csq) else (eval alt)
  | Lang.TST (Lang.Eqn (lhs, rhs), csq, alt) ->
    let lhs' = eval lhs in let rhs' = eval rhs in
      if (lhs' = rhs') then (eval csq) else (eval alt)
  | _ -> raise (RuntimeError "not a number.")
;;

let typing e =
  let rec run e = match e with
    | Lang.Lit _ -> Lang.Int
    | _ -> raise (RuntimeError "not implemented yet.") (* TODO *)
    in let res = run e in (match res with
      | Lang.Int -> ()
      | _ -> raise (TypeError (Lang.Int, res)));;

File: src/main/kotlin/stl/Interpreter.kt
package stl

import java.lang.Exception

fun eval(expr: Expr): Expr =
  when(expr) {
    is Add -> {
      val lhs = eval(expr.lhs)
      val rhs = eval(expr.rhs)
      if (lhs is Lit && rhs is Lit) {
        Lit(lhs.value + rhs.value)
      }
      else {
        throw Exception("add is for int only")
      }
    }
    is App -> {
      val f = eval(expr.f)
      val a = eval(expr.a)
      if (f is Fun) {
        eval(f.body.subst(f.arg, a))
      }
      else {
        throw Exception("not a function")
      }
    }
    is Tst -> {
      val cond = when (expr.cond) {
        is Eqn -> {
          val lhs = eval(expr.cond.lhs)
          val rhs = eval(expr.cond.rhs)
          if (lhs is Lit && rhs is Lit) {
            lhs.value == rhs.value
          }
          else {
            throw Exception("eqn is for int only")
          }
        }
        is Les -> {
          val lhs = eval(expr.cond.lhs)
          val rhs = eval(expr.cond.rhs)
          if (lhs is Lit && rhs is Lit) {
            lhs.value < rhs.value
          }
          else {
            throw Exception("les is for int only")
          }
        }
        else -> throw Exception("not a boolean value")
      }

      if (cond) eval(expr.csq) else eval(expr.alt)
    }
    is Var -> throw Exception("variable ${expr.name} is unbound")
    else -> expr
  }
